### PR TITLE
!!![TASK] Refactor reference handling

### PR DESCRIPTION
**Pull request by linawolf**

### CODE REVIEW

**Comment by linawolf** [comment]
@greg0ire Not sure if I should squash the commits more or keep them separate like this

**Comment by greg0ire** [comment]
Does the "to squash or not to squash" of [our guide](https://www.doctrine-project.org/contribute/index.html) help?

File: docs/en/custom-directives.rst [hunk 1]
```diff
@@ -52,17 +52,19 @@ Example Directive
...
     }
 
-Now you can register your directive by passing it to the 2nd argument of the ``Doctrine\RST\Kernel`` class:
+Now you can register your directive by registering it in a custom directive
+factory in your :file:`Configuration.php`:
 
 .. code-block:: php
 
-    use App\RST\Directives\ExampleDirective;
+    $configuration       = new Configuration();
```
Comment (greg0ire) [inline docs/en/custom-directives.rst]:
```suggestion
    $configuration = new Configuration();
```

File: docs/en/custom-text-roles.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+.. code-block:: rst
+
+    This is some :role:`interpreted text`.
+
+You can define custom text roles in your project if needed:
+
+Text role class
+---------------
+
+You can write your own text coles by defining a class that implements
```
Comment (greg0ire) [inline docs/en/custom-text-roles.rst]:
```suggestion
You can write your own text roles by defining a class that implements
```

File: docs/en/custom-text-roles.rst [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            return '<samp>' . $spanToken->get('text') . '</samp>';
+        }
+    }
+
+Now you can register your text role by registering it in a custom directive
+factory in your :file:`Configuration.php`:
+
+.. code-block:: php
+
+    $configuration       = new Configuration();
```
Comment (greg0ire) [inline docs/en/custom-text-roles.rst]:
```suggestion
    $configuration = new Configuration();
```

File: lib/Environment.php [hunk 4]
```diff
@@ -480,4 +423,41 @@ private function addMissingTextRoleSectionError(string $section, string $type):
...
+    }
+
+    public function resolveDependency(string $data, ResolvedReference $resolvedReference): void
+    {
+        if (! isset($this->unresolvedDependencies[$data]) || $this->getMetaEntry() === null) {
+            return;
+        }
+
+        $this->getMetaEntry()->resolveDependency(
+        // use the unique, unresolved name
```
Comment (greg0ire) [inline lib/Environment.php]:
```suggestion
            // use the unique, unresolved name
```

File: lib/HTML/HTMLFormat.php [hunk 5]
```diff
@@ -21,10 +22,18 @@ final class HTMLFormat implements Format
...
+    /** @var Renderers\LinkRendererFactory[]  */
+    private array $rendererFactories;
+
     public function __construct(TemplateRenderer $templateRenderer)
     {
-        $this->templateRenderer = $templateRenderer;
-        $this->directiveFactory = new FormatDirectiveFactory();
+        $this->templateRenderer  = $templateRenderer;
+        $this->directiveFactory  = new FormatDirectiveFactory();
+        $this->templateRenderer  = $templateRenderer;
```
Comment (greg0ire) [inline lib/HTML/HTMLFormat.php]: :thinking: there is the same line above
```suggestion
```

File: lib/LaTeX/LaTeXFormat.php [hunk 6]
```diff
@@ -12,19 +12,27 @@
...
+
     private DirectiveFactory $directiveFactory;
 
     public function __construct(TemplateRenderer $templateRenderer)
     {
-        $this->templateRenderer = $templateRenderer;
-        $this->directiveFactory = new FormatDirectiveFactory();
+        $this->templateRenderer  = $templateRenderer;
+        $this->directiveFactory  = new FormatDirectiveFactory();
+        $this->templateRenderer  = $templateRenderer;
```
Comment (greg0ire) [inline lib/LaTeX/LaTeXFormat.php]:
```suggestion
```

Comment (linawolf) [reply lib/LaTeX/LaTeXFormat.php]: Must be a merging artifact

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @linawolf !

**Pull request merged**